feat(vis-type: bar): switch to ECharts bar chart

dv-usama-ansari opened the pull request:
Closes https://github.com/datavisyn/visyn_core/issues/89
Follow up of #438 

### Developer Checklist (Definition of Done)

**Issue**

- [ ] All acceptance criteria from the issue are met
- [x] Tested in latest Chrome/Firefox

**UI/UX/Vis**

- [x] Requires UI/UX/Vis review
  - [x] Reviewer(s) are notified @thinkh @dvmoritzschoefl 
  - [ ] Review has occurred (_link to notes_)
  - [ ] Feedback is included in this PR
  - [ ] Reviewer(s) approve of concept and design

**Code**

- [x] Branch is up-to-date with the branch to be merged with, i.e., develop
- [x] Code is cleaned up and formatted
- [ ] Unit tests are written (frontend/backend if applicable)
- [ ] Integration tests are written (if applicable)

**PR**

- [x] Descriptive title for this pull request is provided (will be used for release notes later)
- [x] Reviewer and assignees are defined
- [x] Add type label (e.g., *bug*, *feature*) to this pull request
- [x] Add release label (e.g., `release: minor`) to this PR following [semver](https://semver.org/)
- [x] The PR is connected to the corresponding issue (via `Closes #...`)
- [x] [Summary of changes](#summary-of-changes) is written


### Summary of changes
- Fix bar plot selection

### Screenshots
- TBD

### Additional notes for the reviewer(s)
- N/A

dvdanielamoitzi commented:
Here's some feedback from my testing in visyn_core directly:

- [ ] Unknown category should always be the last one
- [ ] we should hide the labels when the bar is too small in a grouped / stacked bar, as the labels overflow anyways:
![image](https://github.com/user-attachments/assets/6e00
![image](https://github.com/user-attachments/assets/d43c86ca-7710-4e86-b48f-6c74839b6776)
2b51-3227-4bde-a88f-2ec998a4dab3)

- [ ] when removing facets the remaining chart should use 100% of the height again. Currently, there's a lot of whitespace:
![image](https://github.com/user-attachments/assets/225571d4-91a1-4120-9366-50c1393c83d1)

- [ ] Bars are overflowing in stacked mode:
![image](https://github.com/user-attachments/assets/276635c1-42da-4cb0-a3ce-57456321b252)

- [ ] I'd reduce the max length of the tick label by half, as it is very long currently and takes up a lot of space:
![image](https://github.com/user-attachments/assets/c2bb00f8-7753-45c1-8963-13cfae55523c)

- [ ] half of the axis labels is missing (I have the same config as in the above screenshot, except for the bar direction):
![image](https://github.com/user-attachments/assets/688e0f01-3dac-4efa-b4ab-0bab43e52968)

- [ ] Axis labels are missing

- [ ] In the old barchart we could sort the individual axis either by category, or by the count which is not possible currently, but needed from our side:
![image](https://github.com/user-attachments/assets/50dbca76-f3a7-4fea-9b17-58c9b6339698)

- [ ] Aggregate type minimum seems to be wrong imo
![image](https://github.com/user-attachments/assets/f8ab32bd-a3ef-4526-9415-684e6eb33209)

dvmoritzschoefl reviewed (changes requested):
I found some usability issues:

- [ ] The scrollbar is not visible for me. This is probably because of a width/overflow problem. (setting the width manually through the console reveals the scrollbar)
- [ ] Downloading the visualization only shows the currently visible bar charts, and not the whole visualization (this might be ok however since then its WYSIWYG). Also weird for the downloaded image the scrollbars are now visible but without mantine styling
- [ ] When changing the categorical column something weird happens: First a resize is triggered which rerenders the visualization, causing it to "jump". Then the mantine tooltip of the menu item will be on the left top corner of the viewport for a short period of time and then the visualization will finally update.
- [ ] Unknown is always the first category (should it be the last instead?)
- [ ] When hovering the selected bar, I observed different behavior than for all other bars: Only the selection is "visible" and not the whole group which is what I would expect (The other bars in the group are instead invisible?)
- [ ] In my testing ECharts had pretty good animations. When switching from count to minimal for instance it seemed to remount the chart instead. Is this intentional?

thinkh commented:
> In my testing ECharts had pretty good animations. When switching from count to minimal for instance it seemed to remount the chart instead. Is this intentional?

I disabled the animations because they were disturbing and sometimes misleading when enabled. Please take a look by enabling this flag:

https://github.com/datavisyn/visyn_core/blob/f6194c7d0243d54b56302df4c8974e33f707ccb3/src/vis/bar/SingleEChartsBarChart.tsx#L362

I also disabled the merging of series because it led to wrong results when switching the grouping and/or category column. 

https://github.com/datavisyn/visyn_core/blob/f6194c7d0243d54b56302df4c8974e33f707ccb3/src/vis/bar/SingleEChartsBarChart.tsx#L571-L573

I'd be happy to use it if it works correctly and improves the user experience.

dv-usama-ansari commented:
All the comments except those marked unchecked are addressed. This PR is the most up to date version as of https://github.com/datavisyn/visyn_core/pull/497/commits/10a45315dabc89012fe04c0ec62f775aa0a67551 and would close the following unmerged PRs:
- #519
- #438
- #330 
- #272

This would also make these refs redundant:
- https://github.com/datavisyn/visyn_core/tree/mh/echarts_hook
- https://github.com/datavisyn/visyn_core/tree/thinkh/refactor-bar-chart-using-echarts
- https://github.com/datavisyn/visyn_core/tree/och/bar-chart-react-window
- https://github.com/datavisyn/visyn_core/tree/ua/89-remove-arquero-from-bar-plot
- https://github.com/datavisyn/visyn_core/tree/ua/89-bar-plot

thinkh commented on src/vis/bar/utils.ts:
```diff
@@ -1,9 +1,11 @@
-import { bin, desc, op } from 'arquero';
+import { bin as aqBin, desc, op } from 'arquero';
+import { bin as d3Bin, extent, max, min } from 'd3v7';
```
Why do we have arquero and d3 bin still here? I think we should cleanup and remove the functions below. Afterwards please remove also the dependency from the package.json.

dv-usama-ansari replied:
I have cleaned up the utils for bar chart in https://github.com/datavisyn/visyn_core/pull/497/commits/2441b4d7706b364cbf62e4494420de49231eea6e and https://github.com/datavisyn/visyn_core/pull/497/commits/15a516e635e0ef48a6c67ea51da6381bdd30d0a8. Unfortunately we cannot remove arquero just yet because it is still used in heatmap aggregations.

thinkh replied:
All right. Thanks.

thinkh commented:
@dv-usama-ansari Currently open tasks:

- [ ] Global min/max for all facets to calculate the scale for all the echart bar charts
- [ ] Check if there is a flag margin/padding of the domain -> echarts should do this automatically according to [these demos](https://echarts.apache.org/handbook/en/how-to/chart-types/bar/basic-bar/#)
- [ ] Clipped tooltip in view -> [set higher z-index for tooltip](https://echarts.apache.org/en/option.html#tooltip.axisPointer.z)
- [ ] Bar background color for grouped bars, add the alpha to the hex value `#RRGGBBAA` instead of `#RRGGBB` + `opacity` (see [example](https://echarts.apache.org/examples/en/editor.html?code=PYBwLglsB2AEC8sDeAoWszGAG0iAXMgL4A0as2ApgOaXQAmhSp5AHgIKsQDOT569AIZhBhANoAiALIwJJWBIAqAV0pyFAdUr11SgBbLdAMQBOEXQGVhl5dAkBdci3QBPTjybPY3Smcq9YMVR0dGhBAFtKQgkrKjl-DBcQKIUAI0ETeJDYIRFxAFZ5ACYABnkAZgA2eQBGMtg64vqAFnsybIgwSnCLMBcqTydHIiA&_source=echarts-doc-preview)) -> the numbers on top of the bar should not be transparent because the `opacity` is not set
- [ ] Sorting of the categorical axis (Daniela's feedback)
- [ ] A single bar chart should use the full height of the container -> add a config flag use full height with default value `false` (Daniela's feedback)

dv-usama-ansari commented:
@thinkh 
I have addressed all the points mentioned by you in the comment

> Make category labels on the y-axis wider (e.g., more than just `REACTOME_...`)

I made this work how we would expect to see the truncation in DOM elements. We now have a constant to define the maximum label width in pixels and the component would handle truncated text automatically:
https://github.com/datavisyn/visyn_core/blob/3f2337e3122afcecf08ba0a5c6d4b5795d729c66/src/vis/bar/constants.ts#L24

As of https://github.com/datavisyn/visyn_core/pull/497/commits/3f2337e3122afcecf08ba0a5c6d4b5795d729c66, I still have to fix the heights of individual charts when the group type is `Grouped` in `Horizontal` direction with facets enabled.

I would document this in detail in this PR very soon.

dv-usama-ansari commented:
It was a tricky implementation but according to https://github.com/datavisyn/visyn_core/pull/497/commits/13b22fc19387103e0069dec8af0ee9e4ec37a7e9, the bar plot labels are now responsive like we saw in the PowerBI dashboard

[responsive-labels-bar-plot.webm](https://github.com/user-attachments/assets/5ff50f80-4f3f-4a47-96e0-ebb770a08805)

dvmartinweigl commented:
- [ ] setting the group to a gene expression (log TPM+1) in bioinsight makes the application crash
![image](https://github.com/user-attachments/assets/fd80f44d-3ef6-4055-bc71-7797a51979e9)
 
- [ ] In our old bar chart we had the possibility to sort the bars having sort icons besides the axis labels. 
![image](https://github.com/user-attachments/assets/be514cb1-826a-4228-baea-5885a39ef292)

- [ ] in the bioinsight use case it would be useful if the plot would also scale vertically. Having facets we will need the scroll area anyways I guess
![image](https://github.com/user-attachments/assets/145a5dec-24df-44fc-bb38-c2b0ec4b00f1)

- [ ] if I reduce the horizontal space to much a lot of labels on the x-axis are missing and everything seems off
![image](https://github.com/user-attachments/assets/443a2db5-c95c-4315-971d-82d7f61088e5)

- [ ] please use the same title font as used in other vis types and center it
![image](https://github.com/user-attachments/assets/65f71f2e-df5f-44a3-b6fe-5e0ef02922de)

- [ ] having facets we should maybe have a combined title of categorical column and facet
![image](https://github.com/user-attachments/assets/deddd658-00a4-41a3-bc97-1a6e6619a1e8)

- [ ] I think we should use a dedicated color for _Unknown_ in all vis types (maybe gray). We should define one in `colors.ts` @dvdanielamoitzi 
![image](https://github.com/user-attachments/assets/3b43ac2c-7774-462e-babb-e32e78da4bd9)

- [ ] when setting to normalized show _count %_ instead of _count_
![image](https://github.com/user-attachments/assets/2f2c90a2-eae5-42b2-9f92-b3c1cb0df03e)

- [ ] I have a pointer for title and axis labels even though there is no action connected. Better switch to cursor

- [ ] truncate the labels for numerical groups
![image](https://github.com/user-attachments/assets/ba7b4c1f-02aa-42a9-8a64-2ed951d2ad29)

dv-usama-ansari commented:
Thanks for your feedback @dvmartinweigl! I have some comments for you regarding the feedback as below:

> setting the group to a gene expression (log TPM+1) in bioinsight makes the application crash

I would need to check this with you since it does not happen in Ordino as well as visyn_core. I am suspecting that it may be an issue with setting ranges for color scale exactly at this location, but we still need to check:
https://github.com/datavisyn/visyn_core/blob/9f0edaa4486e090bb8727cf58a3f7298be844ba4/src/vis/bar/BarChart.tsx#L376


> In our old bar chart we had the possibility to sort the bars having sort icons besides the axis labels.

Unfortunately, the axis titles in eCharts does not support adding icons, we could only add rich text at the best. We should discuss this with @dvdanielamoitzi if we have a better solution.

> in the bioinsight use case it would be useful if the plot would also scale vertically. Having facets we will need the scroll area anyways I guess

The `IBarConfig` has a property `useFullHeight` which you could use in case you require vertical responsiveness. `useFullHeight` uses container's height instead of a static value when there are no facets:

https://github.com/datavisyn/visyn_core/blob/9f0edaa4486e090bb8727cf58a3f7298be844ba4/src/vis/bar/interfaces.ts#L45

https://github.com/datavisyn/visyn_core/blob/9f0edaa4486e090bb8727cf58a3f7298be844ba4/src/vis/bar/BarChart.tsx#L43

> if I reduce the horizontal space to much a lot of labels on the x-axis are missing and everything seems off

I would make it configurable if the bar chart should hide the labels in the axis if they overlap.

I would address all the other points in follow up commits :+1:

dvmartinweigl commented:
- [ ] always show tooltips for rotated labels, for normal ones only show it when truncated
- [ ] add back sorting with text instead of the sorting icons. Extend the label text with the three states (Unsorted) --> (Ascending) --> (Descending). If possible make the ascending/descending state orange colored
- [ ] remove pointer from title if there are no facets
- [ ] check the case of counts with many digits (rotate labels in bars?)
- [ ] increase the spacing between y-label and plot

dv-usama-ansari commented:
Found a bug in reprovisyn with these settings: 
![image](https://github.com/user-attachments/assets/29ea3149-c72d-4fa4-892b-91cea21bddfa)

dv-usama-ansari commented:
> Found a bug in reprovisyn (test_db) with these settings

Fixed in https://github.com/datavisyn/visyn_core/pull/497/commits/d463e0ce6e068c64670ddb94627d02b869f5127e

![image](https://github.com/user-attachments/assets/c38ef69f-2186-4c23-b81c-d4dc9433d1f3)

thinkh commented:
> > Found a bug in reprovisyn (test_db) with these settings
> 
> Fixed in [d463e0c](https://github.com/datavisyn/visyn_core/commit/d463e0ce6e068c64670ddb94627d02b869f5127e)
> 
> ![image](https://private-user-images.githubusercontent.com/115616380/372622587-c38ef69f-2186-4c23-b81c-d4dc9433d1f3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjc4MTIwNzEsIm5iZiI6MTcyNzgxMTc3MSwicGF0aCI6Ii8xMTU2MTYzODAvMzcyNjIyNTg3LWMzOGVmNjlmLTIxODYtNGMyMy1iODFjLWQ0ZGM5NDMzZDFmMy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMDAxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTAwMVQxOTQyNTFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0yZDA2Y2MxZDhjZmRhOGU3ZDRlYzkwZDUwZDViOTYyNjcwNWI1ZDA3Njc2MjgzZTYzZTI0ZDMyNTU4ZmM4Mjk5JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.QvEaK8wj8hR5Fighx1K5gfwT_Gg14_UwY0Qn4es17Oc)

@dv-usama-ansari is it also fixed in the vertical direction? because your first screenshot shows vertical and not horizontal.

dv-usama-ansari commented:
Potential usability issue:
![very-much-groups](https://github.com/user-attachments/assets/c3d25639-d197-4cab-a92f-09a8742043ee)

thinkh commented:
- [ ] Lazy loading echarts in hook

dv-usama-ansari commented:
> @dv-usama-ansari is it also fixed in the vertical direction? because your first screenshot shows vertical and not horizontal.

Yes, this was a miscalculation which exploded the scales. Since the underlying aggregation is fixed, the bar chart displays normally irrespective of the orientation.

dv-usama-ansari commented:
@thinkh When you are reviewing this, please make sure to check these commits:
- https://github.com/datavisyn/visyn_core/pull/497/commits/b04a18a5a737bb6af8c4ae1976d73c9016533bc8
- https://github.com/datavisyn/visyn_core/pull/497/commits/99de3a612c529d097c0ed7bb8d89cfbfbfdab648
- https://github.com/datavisyn/visyn_core/pull/497/commits/94752fd9973fe2a57035548eb9bec621e3b5d6cd
- https://github.com/datavisyn/visyn_core/pull/497/commits/1090cdb357f5721c1dc461b8ff9420add20175e1: Please also check the `Intl.NumberFormat` here

dvchristianbors commented:
Thanks for your changes. Both the horizontal and vertical bar plots now work much better. I have a few comments that I will put here to be addressed:

![image](https://github.com/user-attachments/assets/35b9d0c8-d904-48b4-ba9e-cf21b2c642c7)
Please add the selected column to the value-axis, otherwise, users won't know what average is used.

dvchristianbors commented:
When the bar plot is scrollable, and neither the title, nor the axis are visible, the user will not know what values they are looking at without adding this information to the tooltip.
![image](https://github.com/user-attachments/assets/a00a9e08-f3b3-4b37-b5df-e9270f2ac82e)

Please again account for different conditions between count and other aggregations

dvchristianbors commented:
Scrollbar and axis title are overlapping
![image](https://github.com/user-attachments/assets/1263cb8f-086f-422b-941d-be8748688a20)

dvchristianbors commented:
Legend cannot be interacted with due to scrollbar
![image](https://github.com/user-attachments/assets/3d67316c-0ffc-43dc-8ad1-4a6b08d5eabf)

dvchristianbors commented:
Please ensure that category values are always used as strings, not numbers
![image](https://github.com/user-attachments/assets/5883e64d-1846-49f3-adc5-a4ed07bf6fbe)

dv-usama-ansari commented:
Implementation with the new tooltip:

![image](https://github.com/user-attachments/assets/6f6bd9b0-1ab9-4f68-b070-2d0e1d3cea55)

> Please ensure that category values are always used as strings, not numbers

This was not reproducible :point_up:

mstreit commented:
> Implementation with the new tooltip:
> 
> ![image](https://private-user-images.githubusercontent.com/115616380/372952741-6f6bd9b0-1ab9-4f68-b070-2d0e1d3cea55.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjc4OTk2MjYsIm5iZiI6MTcyNzg5OTMyNiwicGF0aCI6Ii8xMTU2MTYzODAvMzcyOTUyNzQxLTZmNmJkOWIwLTFhYjktNGY2OC1iMDcwLTJkMGUxZDNjZWE1NS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMDAyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTAwMlQyMDAyMDZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0wMGJmOTkwY2IyYzMyNmVkOGFlNTE5YWY0ZGM4MzRjN2JiNTBlZjY2YTk5MDI2ZjZjMWJkNjA0YzljNzM1MjAxJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.DZPXE22-kwDeXplzcEJOy17DKLHoekGIODCmEIPs-dU)
> 
> > Please ensure that category values are always used as strings, not numbers
> 
> This was not reproducible ☝️

General comment (though probably unrelated to this issue): I would assume that the values shown within the bars are hidden in cases where the bar sections are too small to show values without heavy overlapping. I think this is also the behavior in other off-the-shelf tools like PowerBI. Ordino users should not be able to configure plots with heavy overplotting of labels or other nonsense visualizations (like the one shown in the screenshot). If the user wants to see the exact values of small bar sections, they can hover over the bar to read the value from the tooltip.

dv-usama-ansari commented:
> General comment (though probably unrelated to this issue): I would assume that the values shown within the bars are hidden in cases where the bar sections are too small to show values without heavy overlapping. I think this is also the behavior in other off-the-shelf tools like PowerBI. Ordino users should not be able to configure plots with heavy overplotting of labels or other nonsense visualizations (like the one shown in the screenshot). If the user wants to see the exact values of small bar sections, they can hover over the bar to read the value from the tooltip.

Thanks for your comment @mstreit! I have already taken care of overlapping and echarts provides a flag out of the box to avoid such a condition. The only reason that this plot looks like this is because the values associated with an individual bar is in millions.  (For example: The wider red bar in chromosome 18 has a value of about 75M and due to this, all the other labels are hidden since they would overlap).

Do you suggest that I use number formatter to show them as `75M` instead of `75086514` in the bar plot?

Unfortunately, we do not have a control over the process of hiding the labels since it is done by echarts internally. This is the reason why the bar labels overlap the axis labels in some cases.

mstreit commented:
> > General comment (though probably unrelated to this issue): I would assume that the values shown within the bars are hidden in cases where the bar sections are too small to show values without heavy overlapping. I think this is also the behavior in other off-the-shelf tools like PowerBI. Ordino users should not be able to configure plots with heavy overplotting of labels or other nonsense visualizations (like the one shown in the screenshot). If the user wants to see the exact values of small bar sections, they can hover over the bar to read the value from the tooltip.
> 
> Thanks for your comment @mstreit! I have already taken care of overlapping and echarts provides a flag out of the box to avoid such a condition. The only reason that this plot looks like this is because the values associated with an individual bar is in millions. (For example: The wider red bar in chromosome 18 has a value of about 75M and due to this, all the other labels are hidden since they would overlap). Any bar which is hovered would show its value even if it overlaps the others.
> 
> Do you suggest that I use number formatter to show them as `75M` instead of `75086514` in the bar plot? We already have the whole number displayed in the tooltip.
> 
> Unfortunately, we do not have a control over the process of hiding the labels since it is done by echarts internally. This is the reason why the bar labels overlap the axis labels in some cases.

I see. Thanks for clarifying this. Regarding how it should behave in corner cases, I suggest looking at general-purpose tools like PowerBI or Tableau.

dv-usama-ansari commented:
> ![image](https://private-user-images.githubusercontent.com/101096478/373535467-33f5193e-90bc-4b4d-8c69-b19a0da1e233.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjgwMjY4NjEsIm5iZiI6MTcyODAyNjU2MSwicGF0aCI6Ii8xMDEwOTY0NzgvMzczNTM1NDY3LTMzZjUxOTNlLTkwYmMtNGI0ZC04YzY5LWIxOWEwZGExZTIzMy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMDA0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTAwNFQwNzIyNDFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT00ZDE3MGZkZmMxYjFiNWM3Y2E3MmJlZWIzMWYwZWI2ZjczY2MyZTg5YzE1YWQ0ZGZjZWQ1ODhjYTA5ODY1YWM1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.Pf4oBeVuzTfFv_HsODK1JzKxTHQNrkkQU4b9dECcMnA)

@dvmartinweigl: I have set a minimum width to the bars so that they do not get very narrow. In case when the bar plot does not fir horizontally, a scrollbar is available. Could you please check that? Alternately, I could also add a config to shrink responsively and not use the min-width. Please let me know if I should go ahead with this.

/cc @thinkh

thinkh commented:
> Alternately, I could also add a config to shrink responsively and not use the min-width. Please let me know if I should go ahead with this.

@dv-usama-ansari We could use a similar config like `useFullHeight` for the width as well. Then BioInsight could use the responsive mode that has no minimum bar width.

What do you think @dv-usama-ansari and @dvmartinweigl?

dvmartinweigl commented:
Just tested this again in bioinsight and everything looks good from my side :+1:

thinkh commented:
@dv-usama-ansari the list of feedback discussed in our review:

- [ ] Add unit tests for `/interfaces/internal/helpers/*`
- [ ] Remove sort buttons from header
- [ ] Unknown color in legend and tooltip

![image](https://github.com/user-attachments/assets/a0709245-3bd6-468b-9a96-50ed4c96a2f8)

dv-usama-ansari commented:
> * [x]  Unknown color in legend and tooltip

This has been fixed in https://github.com/datavisyn/visyn_core/pull/497/commits/3bc84de9315f56273d2c949194b08ba79ef9b4c9
![image](https://github.com/user-attachments/assets/b930a029-0890-46cc-ad5d-01d8802a82a2)

dv-usama-ansari commented on src/vis/bar/interfaces/internal/constants/group.ts:
@thinkh This file contains a non-UTF character ` ` therefore it is treated as binary.

thinkh replied:
Can we remove this character?

dv-usama-ansari commented:
Fixed these issues from https://github.com/datavisyn/visyn_core/pull/497/commits/ad99544566b3730ce1a118f57d99d4f60253a193 to https://github.com/datavisyn/visyn_core/pull/497/commits/b9d4752968a446296f1c59096fed5a98a69af2e7:
- https://github.com/datavisyn/visyn_core/issues/563#issuecomment-2404850466
- https://github.com/datavisyn/visyn_core/issues/563#issuecomment-2404852492
- https://github.com/datavisyn/visyn_core/issues/563#issuecomment-2404854719
- https://github.com/datavisyn/visyn_core/issues/563#issuecomment-2404857865
- https://github.com/datavisyn/visyn_core/issues/563#issuecomment-2404874012
- https://github.com/datavisyn/visyn_core/issues/563#issuecomment-2404875704
- https://github.com/datavisyn/visyn_core/issues/563#issuecomment-2404877637
- https://github.com/datavisyn/visyn_core/issues/563#issuecomment-2404879739
- https://github.com/datavisyn/visyn_core/issues/563#issuecomment-2404956316

thinkh commented on src/demo/MainApp.tsx:
```diff
@@ -5,52 +5,66 @@ import { DatavisynTaggle, VisynRanking, autosizeWithSMILESColumn } from '../rank
 import { fetchBreastCancerData } from '../vis/stories/fetchBreastCancerData';
-import { MyCategoricalScore, MyLinkScore, MyNumberScore, MySMILESScore, MyStringScore } from './scoresUtils';
+
+const testData = generateTestData(1000);
```
Unused. Can we remove this?

thinkh reviewed (changes requested):
Thanks for the adding additional test cases today. I checked them on a sample basis and reviewed the remaining files once more. I found some more points with the most critical currently being the random dataset in the storybook stories that will break our visual regression testing with Chromatic. Please find my detailed comments below.

dv-usama-ansari commented on src/vis/bar/interfaces/internal/constants/facet.ts:
```diff
@@ -0,0 +1,4 @@
+/**
+ * Name of default facet when no facet is selected
+ */
+export const DEFAULT_FACET_NAME = '$default$';
```
`$default$` is used as a magic variable, but it is a string value is immutable. I have used this special string to be used as a facet name in the data structure of `aggregatedData` when the facet column is `null`.

dv-usama-ansari commented on src/vis/bar/interfaces/internal/helpers/create-bin-lookup.test.ts:
```diff
@@ -0,0 +1,85 @@
+  it('should return correct number bins for high precision floating point numbers', () => {
+    const binLookup = createBinLookup(
+      Array.from({ length: 10 }, (_, i) => {
+        const val = RNG(i * 100, 'mixed')() * 100;
```
The `RNG` function always returns the same random values when invoked. I have tested it multiple times and it works as expected.

dv-usama-ansari commented:
Storybook fixed in https://github.com/datavisyn/visyn_core/pull/497/commits/45f3fb01dd5b9c8f8510f0fabc046ba90783e598
Below tests are still pending to write (in the same priority order):
- https://github.com/datavisyn/visyn_core/blob/ua/bar-chart-selection/src/vis/bar/interfaces/internal/helpers/sort-series.test.ts
- https://github.com/datavisyn/visyn_core/blob/ua/bar-chart-selection/src/vis/bar/interfaces/internal/helpers/get-bar-data.test.ts
- https://github.com/datavisyn/visyn_core/blob/ua/bar-chart-selection/src/vis/bar/interfaces/internal/helpers/normalized-value.test.ts

I would write tests after finishing up https://github.com/datavisyn/reprovisyn/pull/2059/

dv-usama-ansari commented:
> Grouping by numerical column shows the bins correctly, but the colors are wrongly assigned.
>  
> I'd expect that the colors are assigned from light blue for the lowest values to dark blue for the highest values.

Fixed in https://github.com/datavisyn/visyn_core/pull/497/commits/94862478e95da4cb43ba15ab3a5df145958bb65f
![image](https://github.com/user-attachments/assets/0e041374-e884-45a5-982e-d6690f34c7ef)

thinkh commented:
@dv-usama-ansari Thanks for the fix. I can confirm that it works now as expected.

![image](https://github.com/user-attachments/assets/6653a3ee-1bbf-4e73-8d80-57d26ed76a3a)

thinkh reviewed (dismissed):
@dv-usama-ansari Thanks for the whole implementation and adding tests.  From my point we are ready to merge now. There are still some problems that have been documented in the follow-up issue.

The most recent one I've found: https://github.com/datavisyn/visyn_core/issues/563#issuecomment-2416583243

thinkh commented:
@dv-usama-ansari I saw that the Playwright tests are failing for this branch. in comparison the develop branch succeeded with the latest merge. Please check the tests and fix them. Thanks.

dv-usama-ansari commented:
I have updated the `MainApp` like it was in develop. All of the tests pass except the download tests which are unrelated to this PR: https://github.com/datavisyn/visyn_core/actions/runs/11368336264/job/31623179398

thinkh commented:
Thank you for taking care of this.
